edflow/iterators/tf_trainer.py

```diff
@@ -316,3 +316,100 @@ def run(self, fetches, feed_dict):
+          3:
+            name: train
+            losses: ["model"]
+
+    The stages are sorted by their key. It is recommended to keep the simple numeric ordering.
+    In each stage, a set of losses can be specified through the `losses : [ "loss1", "loss2", ...]` syntax.
+    The duration of each stage is given by the `end : num_steps` value. Note that the end of a stage
+    is determined in the order of the stages. A later stage has to have a higher end value then the previous one.
+
+    The model has to implement the `edflowiterators.tf_trainer.TFMultiStageModel` interface. Look at the multistage_trainer example.
+
+    # TODO: actually, we do not sanity check that the end value of stages make sense with the ordering. This might be nice
+    """
+
+    def __init__(self, config, root, model, **kwargs):
+        if not isinstance(model, TFMultiStageModel):
+            raise ValueError("Model does not support Multistage interface")
+        else:
+            super(TFMultiStageTrainer, self).__init__(config, root, model, **kwargs)
+
+    def create_train_op(self):
+        super(TFMultiStageTrainer, self).create_train_op()
+        self.stages = copy.deepcopy(self.config.get("stages"))
+        for stage_number, stage_options in self.stages.items():
+            stage_opt_ops = [
+                self.opt_ops[loss_name] for loss_name in stage_options["losses"]
+            ]
+            stage_opt_ops = tf.group(*stage_opt_ops)
+            with tf.control_dependencies([stage_opt_ops] + self.update_ops):
+                stage_train_op = tf.no_op()
+            stage_options.update({"train_op": stage_train_op})
+
+    def run(self, fetches, feed_dict):
+        current_stage = self.determine_current_stage()
+        self.session.run(
+            [self.model.stage_update_op],
+            feed_dict={self.model.stage_placeholder: current_stage},
+        )
+        current_train_op = self.get_current_train_op(current_stage)
+        if self.get_global_step() == 0:
+            self.logger.info("Running custom initialization op.")
+            fetches["step_ops"] = self.run_once_op
+        else:
+            self.logger.info("Running training stage : {}".format(current_stage))
+            fetches["step_ops"] = current_train_op
+        return super(TFMultiStageTrainer, self).run(fetches, feed_dict)
+
+    def get_current_train_op(self, current_stage):
+        current_train_op = self.stages[current_stage]["train_op"]
+        return current_train_op
+
+    def determine_current_stage(self):
+        global_step = self.get_global_step()
+        for stage_number in sorted(self.stages.keys()):
+            stage_options = self.stages[stage_number]
+            end = stage_options.get("end", -1)
+            if end < 0:
+                return stage_number
+            elif global_step < end:
+                return stage_number
+            else:
+                pass
+
+
+class TFMultiStageModel(object):
+    def define_graph(self):
```

Comment:
I am not sure if it is really useful to override `define_graph`. Maybe `setup_stages` is a better idea. The child class will have to call the parent anyway.

Reply:
why not just put it into `__init__`?